In [2]:
import pandas as pd
import numpy as np

##### ***For downloading a file use urlretrieve***

In [3]:
from urllib.request import urlretrieve

In [4]:
italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

urlretrieve(italy_covid_url,'italy-covid-daywise.csv')


('italy-covid-daywise.csv', <http.client.HTTPMessage at 0x278688b7650>)

In [5]:
import matplotlib 

In [6]:
import matplotlib.pyplot as plt

In [7]:
import seaborn as sns

In [8]:
covid_df = pd.read_csv('italy-covid-daywise.csv')

##### ***We use 'pd' as pandas function is being on that dataframe***

*The .read_csv is specifically for csv files to read and convert it to pandas dataframe*

In [9]:
type(covid_df)

pandas.core.frame.DataFrame

In [10]:
covid_df

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


***dataframe.info() results column wise details such as datatype and count***

In [11]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [12]:
covid_df.describe()

,new_cases,new_deaths,new_tests
count,248.000000,248.000000,135.000000
mean,1094.818548,143.133065,31699.674074
std,1554.508002,227.105538,11622.209757
min,-148.000000,-31.000000,7841.000000
25%,123.000000,3.000000,25259.000000
50%,342.000000,17.000000,29545.000000
75%,1371.750000,175.250000,37711.000000
max,6557.000000,971.000000,95273.000000


*Column property (arrtibute)of dataframe gives list of column in dataframe*

In [13]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [14]:
covid_df.shape

(248, 4)

*we can read multiple different format by using read.csv()*

In [15]:
#Pandas Dataframe is similar to this

covid_data_dict = {
    'date' : ['date1','date2','date3','date4','date5','date6',],
    'new_cases': [1444,1365,996,975,1326],
    'new_deaths' : [1,4,6,8,6],
    'new_tests' : [5443,653,2432,6342]
}


##### *Imagine this picture when ever you are going to use pandas_dataframe*

##### ***Retrive Data From Dataframe***

In [16]:
covid_df

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


In [17]:
covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

In [18]:
covid_df

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


In [19]:
covid_df['new_tests']

0          NaN
1          NaN
2          NaN
3          NaN
4          NaN
        ...   
243    53541.0
244    42583.0
245    54395.0
246        NaN
247        NaN
Name: new_tests, Length: 248, dtype: float64

***Each Column is Represented by data structure called Series.***

In [20]:
type(covid_df['new_cases'])

pandas.core.series.Series

***We Can Access the value in particular column by using 2d array indexing methods***

In [21]:
covid_df['new_cases'][142]

665.0

*Pandas also provide the .at method to dirrectly retirve at specific row & column*

In [22]:
covid_df.at[142,'new_cases']

665.0

In [23]:
covid_df.new_cases

# Same as covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

*We can access some particular columns by creating a dataframe of that particular columns*

In [24]:
cases_df = covid_df[['date','new_cases']]

In [25]:
cases_df
 

,date,new_cases
0,2019-12-31,0.0
1,2020-01-01,0.0
2,2020-01-02,0.0
3,2020-01-03,0.0
4,2020-01-04,0.0
...,...,...
243,2020-08-30,1444.0
244,2020-08-31,1365.0
245,2020-09-01,996.0
246,2020-09-02,975.0


*All the accessing operation result a dataframe pointing to original dataframe so be cautions around using part of dataframe it can alter original dataframe*

*To Over come this situation we can use .copy() method to make a view of that particular columns of that dataframe*

In [26]:
cases_df2 = covid_df.copy()
cases_df2

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


##### *.loc[index] is used to retrive particular row or multiple rows from dataframe*

In [27]:
covid_df.loc[142]

date          2020-05-21
new_cases          665.0
new_deaths         161.0
new_tests        40644.0
Name: 142, dtype: object

*We Can view few rows from dataframe by using .head(no_of_rows) and .tail(no_of_rows)*

In [28]:
covid_df.head(5)

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN


In [29]:
covid_df.tail(3)

,date,new_cases,new_deaths,new_tests
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN
247,2020-09-03,1326.0,6.0,NaN


In [30]:
covid_df.at[0,'new_tests']

nan

Italy Started sharing new_tests data after 19th april 2020 till then 935310 test have conducted

*We can retrive the index of the row from where the number is started where before it the data is represented in form of NaN*

In [31]:
covid_df.new_tests.first_valid_index()

111

##### ***We Can Also use python slicing to slice between rows***

In [32]:
covid_df[31:61]

,date,new_cases,new_deaths,new_tests
31,2020-01-31,3.0,0.0,NaN
32,2020-02-01,0.0,0.0,NaN
33,2020-02-02,0.0,0.0,NaN
34,2020-02-03,0.0,0.0,NaN
35,2020-02-04,0.0,0.0,NaN
36,2020-02-05,0.0,0.0,NaN
37,2020-02-06,0.0,0.0,NaN
38,2020-02-07,0.0,0.0,NaN
39,2020-02-08,0.0,0.0,NaN
40,2020-02-09,0.0,0.0,NaN


In [33]:
covid_df.sample(25)

,date,new_cases,new_deaths,new_tests
246,2020-09-02,975.0,8.0,NaN
135,2020-05-14,888.0,195.0,41131.0
69,2020-03-09,1492.0,133.0,NaN
86,2020-03-26,5210.0,685.0,NaN
187,2020-07-05,235.0,21.0,21166.0
191,2020-07-09,193.0,15.0,29947.0
186,2020-07-04,223.0,15.0,28946.0
62,2020-03-02,561.0,6.0,NaN
167,2020-06-15,338.0,44.0,17463.0
149,2020-05-28,584.0,117.0,39838.0


##### ***Summary***

 1:- covid_df['column name'] -> access column as series using column name

 2:- new_cases[243] -> access particular row using row index

 3:- covid_df.at[243,'new_cases'] ->retriving a single value from the data frame

 4:- covid_df.copy() -> Makes a copy of dataframe which we not affect the orignal dataframe

 5:-covid_df.loc() -> same as column['row index'] 
            Retriving a row or range of row from dataframe

 6:-covid_df.head(n) -> access first n rows from dataframe

 7:-covid_df.tail(n) -> access last n rows from dataframe

 8:-covid_df.sample(n) -> access random tuple from dataframe to check the distribution of data.

 9:-covid_df.new_tests.first_valid_index -> this is a type of code used for finding first non-empty index in a 'series' 




#### **_________________________________________________________________________**

##### ***Ananlyzing Data from data frames***

Q1.What is the total number of reported cases and deaths related to covid-19 in italy?

In [34]:
total_cases = covid_df['new_cases'].sum()
print(total_cases)

271515.0


In [35]:
total_deaths = covid_df['new_deaths'].sum()
print(total_deaths)

35497.0


In [36]:
death_rate = (total_deaths/total_cases)*100

print("The overall death rate in Italy is {:.2f} %".format(death_rate))

The overall death rate in Italy is 13.07 %


Q2.What is the overall number of test conducted?.A total of 935310 test were conducted before daily test numbers were beign reported.

In [37]:
initial_tests = 935310
total_tests = initial_tests + covid_df['new_tests'].sum()
print("Total Number of tests conducted in italy are {:.2f} ".format(total_tests))


Total Number of tests conducted in italy are 5214766.00 


Q3.What fraction of tests returned a positive result?

In [38]:
tests_positive_rate = total_cases/total_tests
tests_positive_rate
print("Percentage of successful positive test result is {:.2f}%".format(tests_positive_rate*100))

Percentage of successful positive test result is 5.21%


Q11.What percentage of positive test result leads to death 

In [39]:
test_case_death_rate = total_deaths/total_tests

print("ahcbabcu {:.12f}".format((test_case_death_rate)*100))

ahcbabcu 0.680701684409


##### ***Quering and sorting rows***

Q1.Days on whcih number of new cases are more than 1000 ?

In [40]:
MoreThan1000Cases = covid_df.new_cases > 1000

MoreThan1000Cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245    False
246    False
247     True
Name: new_cases, Length: 248, dtype: bool

In [41]:
covid_df.date[MoreThan1000Cases]

68     2020-03-08
69     2020-03-09
70     2020-03-10
72     2020-03-12
73     2020-03-13
          ...    
241    2020-08-28
242    2020-08-29
243    2020-08-30
244    2020-08-31
247    2020-09-03
Name: date, Length: 72, dtype: object

In [42]:
print (covid_df)

           date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]


Q3.Days where ratio of cases reported to test conducted is higher than overall positive 

In [43]:
higher_rate = covid_df[(covid_df.new_cases/covid_df.new_tests) > tests_positive_rate]


In [44]:
higher_rate


,date,new_cases,new_deaths,new_tests
111,2020-04-20,3047.0,433.0,7841.0
112,2020-04-21,2256.0,454.0,28095.0
113,2020-04-22,2729.0,534.0,44248.0
114,2020-04-23,3370.0,437.0,37083.0
116,2020-04-25,3021.0,420.0,38676.0
117,2020-04-26,2357.0,415.0,24113.0
118,2020-04-27,2324.0,260.0,26678.0
120,2020-04-29,2091.0,382.0,38589.0
123,2020-05-02,1965.0,269.0,31231.0
124,2020-05-03,1900.0,474.0,27047.0


In [48]:
covid_df.loc[172]

date          2020-06-20
new_cases         -148.0
new_deaths          47.0
new_tests        29875.0
Name: 172, dtype: object

In [49]:
covid_df_copy = covid_df.copy()

In [50]:
covid_df_copy

,date,new_cases,new_deaths,new_tests
0,2019-12-31,0.0,0.0,NaN
1,2020-01-01,0.0,0.0,NaN
2,2020-01-02,0.0,0.0,NaN
3,2020-01-03,0.0,0.0,NaN
4,2020-01-04,0.0,0.0,NaN
...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0
244,2020-08-31,1365.0,4.0,42583.0
245,2020-09-01,996.0,6.0,54395.0
246,2020-09-02,975.0,8.0,NaN


In [52]:
covid_df_copy.loc[172]

date          2020-06-20
new_cases         -148.0
new_deaths          47.0
new_tests        29875.0
Name: 172, dtype: object

In [57]:
covid_df_copy.loc[168:176]

,date,new_cases,new_deaths,new_tests
168,2020-06-16,301.0,26.0,27762.0
169,2020-06-17,210.0,34.0,33957.0
170,2020-06-18,328.0,43.0,32921.0
171,2020-06-19,331.0,66.0,28570.0
172,2020-06-20,-148.0,47.0,29875.0
173,2020-06-21,264.0,49.0,24581.0
174,2020-06-22,224.0,24.0,16152.0
175,2020-06-23,221.0,23.0,23225.0
176,2020-06-24,113.0,18.0,30237.0


In [ ]:
average = covid_df_copy.new_cases[168:172].sum()
average = average + 148

In [66]:
average

1318.0

In [69]:
covid_df.at[172,'new_cases'] = (covid_df.at[171,'new_cases'] + covid_df.at[173,'new_cases'])/2

297.5

In [71]:
covid_df.loc[168:176]

,date,new_cases,new_deaths,new_tests
168,2020-06-16,301.0,26.0,27762.0
169,2020-06-17,210.0,34.0,33957.0
170,2020-06-18,328.0,43.0,32921.0
171,2020-06-19,331.0,66.0,28570.0
172,2020-06-20,297.5,47.0,29875.0
173,2020-06-21,264.0,49.0,24581.0
174,2020-06-22,224.0,24.0,16152.0
175,2020-06-23,221.0,23.0,23225.0
176,2020-06-24,113.0,18.0,30237.0


____________________________________________________________________________

*Summary*

*covid_df.new_cases.sum() :- finding the sum of values in a column or series*

*covid_df[covid_df.new_cases > 1000] :- querying a subset of rows staisfying the chosen criteria using boolean expressions*

*covid_df.drop('tests_positive_cases')*

*sort*

*covid_df.at[172,'new_cases']*

##### ***Working with dates***

This is time series data

In [73]:
covid_df.date

0      2019-12-31
1      2020-01-01
2      2020-01-02
3      2020-01-03
4      2020-01-04
          ...    
243    2020-08-30
244    2020-08-31
245    2020-09-01
246    2020-09-02
247    2020-09-03
Name: date, Length: 248, dtype: object

In [74]:
covid_df['date'] = pd.to_datetime(covid_df.date)

In [76]:
covid_df.date

0     2019-12-31
1     2020-01-01
2     2020-01-02
3     2020-01-03
4     2020-01-04
         ...    
243   2020-08-30
244   2020-08-31
245   2020-09-01
246   2020-09-02
247   2020-09-03
Name: date, Length: 248, dtype: datetime64[ns]

In [79]:
covid_df['day'] = pd.DatetimeIndex(covid_df.date).day
covid_df['month'] = pd.DatetimeIndex(covid_df.date).month
covid_df['year'] = pd.DatetimeIndex(covid_df.date).year
covid_df['weekday'] = pd.DatetimeIndex(covid_df.date).weekday

In [80]:
covid_df

,date,new_cases,new_deaths,new_tests,day,month,year,weekday
0,2019-12-31,0.0,0.0,NaN,31,12,2019,1
1,2020-01-01,0.0,0.0,NaN,1,1,2020,2
2,2020-01-02,0.0,0.0,NaN,2,1,2020,3
3,2020-01-03,0.0,0.0,NaN,3,1,2020,4
4,2020-01-04,0.0,0.0,NaN,4,1,2020,5
...,...,...,...,...,...,...,...,...
243,2020-08-30,1444.0,1.0,53541.0,30,8,2020,6
244,2020-08-31,1365.0,4.0,42583.0,31,8,2020,0
245,2020-09-01,996.0,6.0,54395.0,1,9,2020,1
246,2020-09-02,975.0,8.0,NaN,2,9,2020,2


In [82]:
covid_df_may_2020 = covid_df[covid_df.month == 5]
covid_df_may_2020

,date,new_cases,new_deaths,new_tests,day,month,year,weekday
122,2020-05-01,1872.0,285.0,43732.0,1,5,2020,4
123,2020-05-02,1965.0,269.0,31231.0,2,5,2020,5
124,2020-05-03,1900.0,474.0,27047.0,3,5,2020,6
125,2020-05-04,1389.0,174.0,22999.0,4,5,2020,0
126,2020-05-05,1221.0,195.0,32211.0,5,5,2020,1
127,2020-05-06,1075.0,236.0,37771.0,6,5,2020,2
128,2020-05-07,1444.0,369.0,13665.0,7,5,2020,3
129,2020-05-08,1401.0,274.0,45428.0,8,5,2020,4
130,2020-05-09,1327.0,243.0,36091.0,9,5,2020,5
131,2020-05-10,1083.0,194.0,31384.0,10,5,2020,6


In [83]:
covid_df_may_2020_matrix = covid_df_may_2020[['new_cases','new_tests','new_deaths']]

In [84]:
covid_df_may_2020_matrix

,new_cases,new_tests,new_deaths
122,1872.0,43732.0,285.0
123,1965.0,31231.0,269.0
124,1900.0,27047.0,474.0
125,1389.0,22999.0,174.0
126,1221.0,32211.0,195.0
127,1075.0,37771.0,236.0
128,1444.0,13665.0,369.0
129,1401.0,45428.0,274.0
130,1327.0,36091.0,243.0
131,1083.0,31384.0,194.0


In [85]:
covid_df_may_2020_matrix.sum()

new_cases       29073.0
new_tests     1078720.0
new_deaths       5658.0
dtype: float64

In [86]:
covid_df.new_cases.mean()

1096.6149193548388

In [91]:
covid_df[covid_df.weekday == 5].new_cases.mean()

1198.5285714285715

***So on sundays the cases where reported more than overall average***